pkg/render/render_test.go

```diff
@@ -0,0 +1,309 @@
+			imagesFile := "not_found"
+			infrastructureFile := "not_found"
+			destination, err := ioutil.TempDir("", "test")
+			assert.NoError(t, err)
+
+			if tc.imagesContent != "" {
+				file, err := ioutil.TempFile(os.TempDir(), configPath)
+				path := file.Name()
+				assert.NoError(t, err)
+				defer file.Close()
+
+				_, err = file.WriteString(tc.imagesContent)
+				assert.NoError(t, err)
+				imagesFile = path
+			}
+
+			if tc.infraContent != "" {
+				file, err := ioutil.TempFile(os.TempDir(), infraPath)
+				path := file.Name()
+				assert.NoError(t, err)
+				defer file.Close()
+
+				_, err = file.WriteString(tc.infraContent)
+				assert.NoError(t, err)
+				infrastructureFile = path
+			}
+
+			r := New(infrastructureFile, imagesFile)
+			err = r.Run(destination)
+			if tc.expectError != "" {
+				assert.EqualError(t, err, tc.expectError)
+				return
+			}
+			assert.NoError(t, err)
+
+			// Assert all files were written to bootstrap dir
+			files, err := ioutil.ReadDir(path.Join(destination, bootstrapPrefix))
+			assert.NoError(t, err)
+			assert.Len(t, files, len(tc.expectObjects))
+		})
+	}
+}
+
+func TestWriteAssets(t *testing.T) {
+	tc := []struct {
+		name          string
+		destination   string
+		preCreateMode fs.FileMode
+		objects       []client.Object
+		expectErr     string
+	}{{
+		name:        "Writing file finished with success",
+		destination: "test",
+		objects:     []client.Object{matchingInfra, matchingConfigMap},
+	}, {
+		name:      "Fail to write into /dev/null",
+		objects:   []client.Object{matchingInfra, matchingConfigMap},
+		expectErr: "mkdir /dev/null: not a directory",
+	}, {
+		name:          "Fail to write into /dev/null",
+		objects:       []client.Object{matchingInfra, matchingConfigMap},
+		destination:   "bad_permissions",
+		preCreateMode: 0444,
+		expectErr:     "permission denied",
+	}}
+
+	for _, tc := range tc {
+		t.Run(tc.name, func(t *testing.T) {
+			destination := "/dev/null"
+
+			if tc.destination != "" {
+				dirPath, err := ioutil.TempDir("", tc.destination)
+				assert.NoError(t, err)
+				destination = dirPath
+				if tc.preCreateMode != 0 {
+					os.MkdirAll(path.Join(destination, bootstrapPrefix), tc.preCreateMode)
+					assert.NoError(t, err)
+				}
+			}
+
+			err := writeAssets(destination, tc.objects)
+			if tc.expectErr != "" {
+				assert.Contains(t, err.Error(), tc.expectErr)
+				return
+			}
+			assert.NoError(t, err)
+
+			// Assert all files were written to bootstrap dir
+			files, err := ioutil.ReadDir(path.Join(destination, bootstrapPrefix))
+			assert.NoError(t, err)
+			assert.Len(t, files, len(tc.objects))
+
+			names := []string{}
+			for _, file := range files {
+				names = append(names, file.Name())
+			}
+			for _, res := range tc.objects {
+				filename := fmt.Sprintf("%s-%s.yaml", res.GetName(), strings.ToLower(res.GetObjectKind().GroupVersionKind().Kind))
+				found := assert.Contains(t, names, filename)
+
+				if found {
```

Comment:
nit: I think it might make sense to split this case into two. This if condition block looks bit strange for me.
Up to you.